### PR TITLE
Revise ExternalTexture Video Test Expectations

### PR DESCRIPTION
**Pull request by bjjones**
Revises external texture video test expectations to test for accurate
color space conversions. Adds BT.709 and BT.2020 video samples.

Issue: #910

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [x] Tests are properly located in the test tree.
- [x] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [x] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [x] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/web_platform/external_texture/video.spec.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ Tests for external textures from HTMLVideoElement (and other video-type sources?
 - videos with various encodings, color spaces, metadata
 
 TODO: consider whether external_texture and copyToTexture video tests should be in the same file
+TODO: add a test configuration for red-green.mp4 to the importExternalTexture Sample test once the infrastructure can load it succesfully.
```
Comment (kainino0x) [inline src/webgpu/web_platform/external_texture/video.spec.ts:7]: What infrastructure can't load it?

If it's in Chromium, we should just add the test and add a suppression in Dawn's `expectations.txt`. (Though this can be a separate PR if it's nontrivial.)

File: src/webgpu/web_platform/external_texture/video.spec.ts [hunk 2]
```diff
@@ -14,11 +15,38 @@ import { startPlayingAndWaitForVideo } from '../../web_platform/util.js';
...
+    video: 'red-green.bt709.vp9.webm',
+    redExpectation: new Uint8Array([0xff, 0x00, 0x00, 0xff]),
+    greenExpectation: new Uint8Array([0x00, 0xff, 0x00, 0xff]),
+  },
+  {
+    video: 'red-green.bt2020.vp9.webm',
+    redExpectation: new Uint8Array([0xff, 0x00, 0x00, 0xff]),
+    greenExpectation: new Uint8Array([0x00, 0xff, 0x00, 0xff]),
+  },
 ];
```
Comment (kainino0x) [inline src/webgpu/web_platform/external_texture/video.spec.ts]: If you use `as const` here you probably won't need to define the `VideoExpectations` type.
```
] as const;
```

File: src/webgpu/web_platform/external_texture/video.spec.ts [hunk 3]
```diff
@@ -99,17 +127,16 @@ g.test('importExternalTexture,sample')
...
-a microtask.
-TODO: Multiplanar scenarios
+supported video formats {vp8, vp9, ogg}, common source colorspaces {bt.601, bt.709, bt.2020}, 
+and to ensure the GPUExternalTexture is destroyed by a microtask.
 `
   )
   .params(u =>
     u //
-      .combine('videoSource', kVideoSources)
+      .combine('videoExpectations', kVideoExpectations)
```
Comment (kainino0x) [inline src/webgpu/web_platform/external_texture/video.spec.ts]: This adds the expectations to the query string, which is unnecessary. You can get them passed through with a "private" param which is a param that starts with `_`. I'll do this in a commit that I'm pushing to your branch.

File: src/webgpu/web_platform/external_texture/video.spec.ts [hunk 4]
```diff
@@ -310,7 +337,7 @@ Tests that we can import an HTMLVideoElement into a GPUExternalTexture and use i
         kFormat,
         { x: 1, y: 0 },
         {
-          exp: new Uint8Array([0x00, 0xff, 0x00, 0xff]),
+          exp: kVideoExpectations[0].greenExpectation,
         }
```
Comment (kainino0x) [inline src/webgpu/web_platform/external_texture/video.spec.ts:340]: FYI, if you run into any issues with precision (needing a threshold because there is a range of possible results): #1121

File: src/webgpu/web_platform/external_texture/video.spec.ts [hunk 5]
```diff
@@ -242,7 +269,7 @@ Tests that we can import an HTMLVideoElement into a GPUExternalTexture and use i
 `
   )
   .fn(async t => {
-    const videoUrl = getResourcePath('red-green.webmvp8.webm');
+    const videoUrl = getResourcePath(kVideoExpectations[0].video);
```
Comment (kainino0x) [inline src/webgpu/web_platform/external_texture/video.spec.ts]: Any reason this test can only run this case and not the others?

**Review by kainino0x — APPROVED** [review]

**Comment by kainino0x** [comment]
I pushed some changes to your branch so you'll need to pull from it before making any further revisions.

File: src/webgpu/web_platform/external_texture/video.spec.ts [hunk 6]
```diff
@@ -99,14 +122,13 @@ g.test('importExternalTexture,sample')
   .desc(
     `
 Tests that we can import an HTMLVideoElement into a GPUExternalTexture, sample from it for all
-supported video formats {vp8, vp9, ogg, mp4}, and ensure the GPUExternalTexture is destroyed by
-a microtask.
-TODO: Multiplanar scenarios
+supported video formats {vp8, vp9, ogg}, common source colorspaces {bt.601, bt.709, bt.2020},
+and to ensure the GPUExternalTexture is destroyed by a microtask.
```
Comment (shaoboyan) [inline src/webgpu/web_platform/external_texture/video.spec.ts:126]: Note that we don't need to ensure GPUExternalTexture is destroyed by a microtask. It is not true for current spec any more.

Comment (kainino0x) [reply src/webgpu/web_platform/external_texture/video.spec.ts:126]: Good catch, I didn't think about this test description very closely.

I assume this test still tests that it's destroyed in a microtask, so please add a TODO here and then we should actually fix the test separately.

Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: This test doesn't actually test the destroy behavior - my guess is that I decided to split the test out in a prior commit. It looks like Shaobo is already revising it, so I'll just remove it from the description and leave the destroy test as is.

File: src/webgpu/web_platform/external_texture/video.spec.ts (hunk 5 above)
Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: I don't believe there's any restriction. I went ahead and added all of them.

File: src/webgpu/web_platform/external_texture/video.spec.ts (hunk 4 above)
Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts:343]: ack. It's consistent for me on my test platform, but I will not be surprised if a tolerance is needed when it is running on all platforms.

File: src/webgpu/web_platform/external_texture/video.spec.ts (hunk 1 above)
Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: Removed the TODO. Not sure if this is a cts issue or a Chromium issue, but I've noted it in a bug here:  https://bugs.chromium.org/p/dawn/issues/detail?id=1432

At the least, running it the standalone cts with Chromium will always fail to load. It hits this error: https://github.com/gpuweb/cts/blob/main/src/webgpu/web_platform/util.ts#L36

It's already disabled in chromium here: https://source.chromium.org/chromium/chromium/src/+/main:content/test/gpu/gpu_tests/test_expectations/webgpu_cts_expectation_queries.js;l=1618

Note that all external texture tests have been disabled in Chromium for awhile for some unknown reason, which I intend to turn back on soon here: https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/webgpu-cts/expectations.txt;l=242?q=expectations.txt

File: src/webgpu/web_platform/external_texture/video.spec.ts (hunk 4 above)
Comment (kainino0x) [reply src/webgpu/web_platform/external_texture/video.spec.ts:343]: Should we dry run this? I forget where the instructions are, but they've changed anyway, so I may as well try it out and document them.

File: src/webgpu/web_platform/external_texture/video.spec.ts (hunk 1 above)
Comment (kainino0x) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: Ah, cool, I see now the mp4 was already there. I think the parameters match now so the existing expectation should still work. (A dry run will double check this, too.)

File: src/webgpu/web_platform/external_texture/video.spec.ts (hunk 4 above)
Comment (kainino0x) [reply src/webgpu/web_platform/external_texture/video.spec.ts:343]: Turns out the dry run process hasn't been simplified the way I thought it might be yet, but here's one with the old method:
https://dawn-review.googlesource.com/c/dawn/+/91760

It has changed slightly so hopefully I got it right.

- check the current CTS revision in Dawn's DEPS file
- rebase the candidate change on that revision
- push to a branch in gpuweb/cts
- change Dawn's DEPS file to the revision of that branch
- wait for the branch to get mirrored to https://chromium.googlesource.com/external/github.com/gpuweb/cts/
- run `gen_ts_dep_lists.py`
- commit and upload a dawn PR
- run `dawn-*-deps-rel` tryjobs

Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts:343]: I'm not sure. I added a Mac tryjob to your CL, which is the one I'm worried about because of this comment: https://source.chromium.org/chromium/chromium/src/+/main:ui/gfx/color_space.h;l=91

I don't know if this enum is even used in the path we're using, and it may not have any impact anyway since we're testing primaries.

Comment (kainino0x) [reply src/webgpu/web_platform/external_texture/video.spec.ts:343]: Thanks. I only didn't add Mac at first because I wanted to make sure it was building at all (which failed the first time around).

That comment sounds like it applies only to the final display of the overlays, and not the data we would get by reading out of textures, but it's vague so hard to be sure. Good to test.

Comment (kainino0x) [reply src/webgpu/web_platform/external_texture/video.spec.ts:343]: (BTW, once those pass, feel free to land this PR.)

Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts:343]: Passed. Merging
